### PR TITLE
Port MASTG test 0039 (by @guardsquare)

### LINKED ISSUE #3014: MASTG v1->v2 MASTG-TEST-0039: Testing whether the App is Debuggable (android)
Create a new MASTG v2 test covering for":

- **Title:** Testing whether the App is Debuggable
- **ID:** MASTG-TEST-0039
- **Link:** https://mas.owasp.org/MASTG/tests/android/MASVS-RESILIENCE/MASTG-TEST-0039/
- **Platform:** android
- **MASVS v1:** ['MSTG-CODE-2']
- **MASVS v2:** ['MASVS-RESILIENCE-4']

**Follow the [guidelines](https://docs.google.com/document/d/1veyzE4cVTSnIsKB1DOPUSMhjXow_MtJOtgHeo5HVoho/edit?usp=sharing)**

### PR DESCRIPTION
**Pull request by titze**
Thank you for submitting a Pull Request to the OWASP MASTG. Please make sure that:

- [x] Your contribution is written in the 2nd person (e.g. you)
- [x] Your contribution is written in an active present form for as much as possible.
- [x] You have made sure that the reference section is up to date (e.g. please add sources you have used, make sure that the references to MITRE/MASVS/etc. are up to date)
- [x] Your contribution has proper formatted markdown and/or code
- [x] Any references to website have been formatted as [TEXT](URL “NAME”)
- [x] You verified/tested the effectiveness of your contribution (e.g.: is the code really an effective remediation? Please verify it works!)

If your PR is related to an issue. Please end your PR test with the following line:
This PR closes #3014.

I did not copy the dynamic part, since this has no benefits over the static analysis (imo).

### CODE REVIEW

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+## Overview
+
+This test case checks if the application has the debuggable flag set in the AndroidManifest.xml. If this flag is set, an attacker can attach a debugger, and e.g., read and modify internals of the application.
+
+!!! note Other ways to debug the application
+    Not enabling debugging in the AndroidManifest.xml does fully prevent all possibilities to debug the app. See @MASWE-0101 for more details on how to prevent debugging.
+
+## Steps
+
+1. View the AndroidManifest.xml, e.g., via @MASTG-TOOL-0121.
```
Comment (TheDauntless) [inline tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md:18]: TOOL-0121 is objdump (ios)

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+---
+
+## Overview
+
+To enable debugging of Webviews, the API `WebView.setWebContentsDebuggingEnabled(true);` can be used to enable WebView debugging (see ["see Android documentation"](https://developer.chrome.com/docs/devtools/remote-debugging/webviews/#configure_webviews_for_debugging)).
+
+Calling this API will enable WebView debugging for **all** of the application's WebViews, independently of the AndroidManifest.xml debuggable flag (See @MASTG-TEST-0x39-1). This allows an attached PC to eavesdrop on, and modify all communication inside WebViews.
+
+!!! note Other ways to debug the application
+    Disabling WebView debugging does not fully prevent all possibilities to debug the app and the WebViews. See @MASWE-0101 for more details on how to prevent debugging.
```
Comment (cpholguera) [inline tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md:16]: I'm not sure if this note will confuse more than help. Let me know what you think: 

What are other ways of debugging the webview if it's not using this? Do you mean that the app can be debugged by other means (but not necessarily the webviews). I guess an attacker able to use frida could call that to enable debugging ...

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+## Steps
+
+1. Reverse engineer the app (@MASTG-TECH-0017).
+2. Inspect all locations where `WebView.setWebContentsDebuggingEnabled` is called, and check
+    1. if it is set to true, and
+    2. if the API call can be executed during runtime.
```
Comment (cpholguera) [inline tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md:23]:
```suggestion
2. Obtain all locations where `WebView.setWebContentsDebuggingEnabled` is called and set to `true`.
```

Comment (cpholguera) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md:23]: what's meant by "if the API call can be executed during runtime"?

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+## Observation
+
+The output should contain all locations where `WebView.setWebContentsDebuggingEnabled` can be called with `true` at runtime.
+
+## Evaluation
+
+The test case fails if any locations in the application can be executed ar runtime, where WebView debugging is enabled.
```
Comment (cpholguera) [inline tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md:31]: same question as above about the "runtime" aspect. Shouldn't we just say that if we found this being `true` that must be fixed?

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+title: Debuggable Flag Not Disabled in the AndroidManifest
+platform: android
+id: MASTG-TEST-0x39-1
+type: [static]
+weakness: MASWE-0067
+---
+
+## Overview
+
+This test case checks if the application has the debuggable flag set in the AndroidManifest.xml. If this flag is set, an attacker can attach a debugger, and e.g., read and modify internals of the application.
```
Comment (cpholguera) [inline tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md:11]:
```suggestion
This test case checks if the application has the debuggable flag ([`android:debuggable`](https://developer.android.com/guide/topics/manifest/application-element#debug)) set to `true` in the AndroidManifest.xml. If this flag is set, an attacker can attach a debugger, and e.g., read and modify internals of the application.

Having this flag set to `true` [is not considered a vulnerability](https://developer.android.com/privacy-and-security/risks/android-debuggable), however, it allow attackers to have more access to the app and resources than intended.
```

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+!!! note Other ways to debug the application
+    Not enabling debugging in the AndroidManifest.xml does fully prevent all possibilities to debug the app. See @MASWE-0101 for more details on how to prevent debugging.
```
Comment (cpholguera) [inline tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md:14]: If we know about other ways to debug the app we should shortly note them here.

Do you mean something like the attacker can set this attribute (which would invalidate the signature) and repackage & re-sign and be able to debug the app. I think this is something we could _note_ here.

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md (hunk 1 above)
Comment (cpholguera) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md:18]: we should use MASTG-TOOL-0123

Comment (cpholguera) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md:18]: But even better, use a technique "Obtaining Information from the AndroidManifest" (or similar title) with 2 sections to 
- get the whole android manifest: can use apktool or jadx/jadx-gui
- or read properties from it: can use aapt

Comment (cpholguera) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md:18]: Always better to use a tech. from a test rather than a tool.

**Review by cpholguera — CHANGES_REQUESTED** [review]

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+id: MASTG-TEST-0x39-2
+type: [static]
+weakness: MASWE-0067
+---
+
+## Overview
+
+To enable debugging of Webviews, the API `WebView.setWebContentsDebuggingEnabled(true);` can be used to enable WebView debugging (see ["see Android documentation"](https://developer.chrome.com/docs/devtools/remote-debugging/webviews/#configure_webviews_for_debugging)).
+
+Calling this API will enable WebView debugging for **all** of the application's WebViews, independently of the AndroidManifest.xml debuggable flag (See @MASTG-TEST-0x39-1). This allows an attached PC to eavesdrop on, and modify all communication inside WebViews.
```
[the remark this answers is not in the document]
Comment (titze) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md:13]: Yes :exploding_head: 

> This setting applies to all of the application's WebViews.

https://developer.chrome.com/docs/devtools/remote-debugging/webviews/#configure_webviews_for_debugging

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md (hunk 2 above)
Comment (titze) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md:16]: I meant that setting debuggable=false still allows someone with root to attach a debugger or attach Frida, essentially debugging the app. This is handled in MASWE-101, not 100% it needs a mention here.

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md (hunk 3 above)
Comment (titze) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md:23]: https://developer.chrome.com/docs/devtools/remote-debugging/webviews/#configure_webviews_for_debugging contains a code snippet that ties the API call to the manifest debuggable flag.

What this sentence should mean is that the API call itself is not a problem if it is essentially in dead code. Maybe it still should not be in the app...
```
if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
    if (0 != (getApplicationInfo().flags & ApplicationInfo.FLAG_DEBUGGABLE))
    { WebView.setWebContentsDebuggingEnabled(true); }
}
```

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md (hunk 6 above)
Comment (titze) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md:16]: I meant https://github.com/OWASP/owasp-mastg/pull/3042#discussion_r1833094787 here as well.

The resigning is a good point, not sure we want to open this topic here.

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md (hunk 1 above)
Comment (titze) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-1.md:20]: @TheDauntless I think this overlapped with landing another tool. Fixed.

@cpholguera I split it up into tech and tool.

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (cpholguera) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md:13]: Oh wow 😨

File: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md (hunk 3 above)
Comment (cpholguera) [reply tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x39-2.md:23]: Ok, got it! Let's see if we should reformulate and/or use this for the evaluation and remediation as well but with more detail mentioning those conditions.

**Review by cpholguera — APPROVED** [review]

**Pull request merged**